### PR TITLE
Scripts: Make React Fast Refresh work with multiple blocks

### PR DESCRIPTION
**Pull request by gziolo**
## What?

Related to https://github.com/WordPress/gutenberg/issues/25188.
Follow-up for https://github.com/WordPress/gutenberg/pull/28273.
Related comments:
- Issue raised in https://github.com/WordPress/gutenberg/issues/25188#issuecomment-1099474694 by @amjard360.
- Issue reported in https://github.com/WordPress/gutenberg/pull/28273#issuecomment-1434552486, and the workaround  shared in https://github.com/WordPress/gutenberg/pull/28273#issuecomment-1445471529 by @rafaucau.
- Follow-up https://github.com/WordPress/gutenberg/pull/28273#issuecomment-1768244883 from me.

`wp-scripts start --hot` works perfectly fine with a single entry point that contains all the logic used in the editor. It starts to break if multiple entry points get loaded in the project editor. A perfect example would be a plugin with numerous blocks where each of them loads JavaScript separately. The workaround so far was to bundle all JavaScript code for the editor together. However, it isn’t always the most efficient approach. Besides, it would be great if that worked out of the box, no matter how configured the entry points were.

Screenshot from https://github.com/WordPress/gutenberg/issues/25188#issuecomment-1104103927:

Screenshot 2022-04-20 at 17 58 19

## Why?

Source: https://github.com/pmmmwh/react-refresh-webpack-plugin/blob/main/docs/TROUBLESHOOTING.md#component-not-updating-with-bundle-splitting-techniques

> Webpack allows various bundle splitting techniques to improve performance and cacheability. However, these techniques often result in a shuffled execution order, which will break fast refresh.
>
> Only one copy of both the HMR runtime and the plugin's runtime should be embedded for one Webpack app
> 
> This concern only applies when you have multiple entry points. You can use Webpack's optimization.runtimeChunk option to enforce this.
> 
> 
> ```js
> module.exports = {
>   optimization: {
>     runtimeChunk: 'single',
>   },
> };
> ```

## How?

In the development mode when React Fast Refresh is enabled, there is a shared runtime created. This way it acts as en entry point for existing scripts that orchestrates all updates with hot-module replacement for React code.

## Known Issues

It looks like CSS support for hot-module reloading doesn't work anymore in `trunk` and with this branch. I will investigate this separately as it is unrelated to these changes. I found an issue that seems very similar https://github.com/webpack-contrib/mini-css-extract-plugin/issues/1089. My debugging session pointed me in the direction that `mini-css-extract-plugin` is unable to correctly match the url of CSS file in the `link` tags with these coming from hot reload. The reason seems to originate from the fact that the asset version appended to the url confuses the plugin.

## Testing Instructions

Create a new test plugin with more than one block. You can do it as follows:

1. Run `npx wp-create-block my-plugin --no-wp-scripts `.
2. Refactor the plugin so it contains more than one block. Using `npx wp-create-block my-block --no-plugin` is helpful here.
3. Go to the plugins folder and run the build in the development mode with HMR/React Fast Refresh enabled:
    ```bash
    cd my-plugin
    ../node_modules/.bin/wp-scripts start --hot
    ```
4. Go to the WordPress instance and activate the test plugin.
5. Edit React components and save changes for every block in the plugin. Observe that React components get updated instantly in the browser without the need to refresh the page.

### CODE REVIEW

File: packages/dependency-extraction-webpack-plugin/lib/index.js [hunk 1]
```diff
@@ -302,6 +302,15 @@ class DependencyExtractionWebpackPlugin {
 				assetData.type = 'module';
 			}
 
+			if ( compilation.options?.optimization?.runtimeChunk !== false ) {
```
Comment (gziolo) [inline packages/dependency-extraction-webpack-plugin/lib/index.js:390]: Should it also be different than `undefined`?

File: packages/dependency-extraction-webpack-plugin/lib/index.js [hunk 2]
```diff
@@ -302,6 +302,15 @@ class DependencyExtractionWebpackPlugin {
...
+				assetData.handle =
+					uniqueName +
```
Comment (gziolo) [inline packages/dependency-extraction-webpack-plugin/lib/index.js]: Interesting, the unique name might not be defined based on the failing unit tests. By the way, this change will most likely require updating test snapshots as sometimes the handle is going to be printed.

File: packages/dependency-extraction-webpack-plugin/lib/index.js (hunk 1 above)
Comment (gziolo) [reply packages/dependency-extraction-webpack-plugin/lib/index.js:390]: The default is `false` as explained:

https://webpack.js.org/configuration/optimization/#optimizationruntimechunk

Other allowed values are `true`, any string and an object.

File: packages/dependency-extraction-webpack-plugin/lib/index.js (hunk 2 above)
Comment (gziolo) [reply packages/dependency-extraction-webpack-plugin/lib/index.js]: Fixed in https://github.com/WordPress/gutenberg/pull/64924/commits/dad5d0482424a7db4d547bf0fae0c05096c3f90b.

File: packages/scripts/config/webpack.config.js [hunk 3]
```diff
@@ -340,6 +341,31 @@ const scriptConfig = {
 								}
 							} );
 
+							if ( hasReactFastRefresh ) {
+								const runtimePath = relative(
```
Comment (gziolo) [inline packages/scripts/config/webpack.config.js:346]:
```suggestion

								// Prepends the file reference to the shared runtime chunk to every script type defined for the block.
								const runtimePath = relative(
```

File: packages/dependency-extraction-webpack-plugin/lib/index.js [hunk 4]
```diff
@@ -387,6 +387,15 @@ class DependencyExtractionWebpackPlugin {
 				assetData.type = 'module';
 			}
 
+			if ( compilation.options?.optimization?.runtimeChunk !== false ) {
+				assetData.handle =
```
Comment (gziolo) [inline packages/dependency-extraction-webpack-plugin/lib/index.js:392]:
```suggestion

				// Sets the script handle for the shared runtime file so WordPress registers it only once when using the asset file.
				assetData.handle =
```

**Review by sirreal — APPROVED** [review]
I tested and I was able to reproduce the lack of refresh using the published wp-scripts package.

When I built using the changes in this PR, refresh worked without issues.

**Comment by ktmn** [comment]
@gziolo Hey did you see the comment: https://github.com/WordPress/gutenberg/commit/fcee058e3bb787ee041c170958e15e66a6302352#r150652268

It's still stopping me from using `--hot` because none of my blocks use `script`, `editorScript` or `viewScript`. 

> Testing Instructions
> Run npx wp-create-block my-plugin --no-wp-scripts .

The `npx wp-create-block` didn't work (outdated?), but I did `npx @wordpress/create-block my-plugin`, edit `block.json` to remove:

```json
"editorScript": "file:./index.js",
"viewScript": "file:./view.js"
```

and then in `package.json` change the script to `"start": "wp-scripts start --hot"`, run, and the result is:

> ERROR in Cannot convert undefined or null to object

With `--stats=verbose` it points to here:

https://github.com/WordPress/gutenberg/blob/b06588e805acbe6ba92124e495055781eb21a7ce/packages/scripts/config/webpack.config.js#L354-L358

This fixes it:

```js
const fields = getBlockJsonScriptFields( blockJson ) || {};
```

---

Also, is it normal that the `*.asset.php` files have `'handle' => 'undefined-[name]'`? 

https://github.com/WordPress/gutenberg/blob/b06588e805acbe6ba92124e495055781eb21a7ce/packages/dependency-extraction-webpack-plugin/lib/index.js#L392-L397

`compilation.name` seems to be undefined.

---

Another change in `@wordpress/scripts` that I must have missed (when upgrading from `30.5.1`) is that I need to enqueue the `runtime.js` file that is generated to `build` dir myself? Or is there some logic that should do it automatically? Either way, looking at the [docs](https://developer.wordpress.org/block-editor/packages/packages-scripts/) there's no mention of "runtime" at all - maybe something that should be documented. 

> --hot – enables “Fast Refresh”. The page will automatically reload if you make changes to the code. For now, it requires that WordPress has the [SCRIPT_DEBUG](https://developer.wordpress.org/advanced-administration/debug/debug-wordpress/#script_debug) flag enabled and the [Gutenberg](https://wordpress.org/plugins/gutenberg/) plugin installed.

Is the Gutenberg part still the case? I usually develop without it and occasionally turn it on to check that what I have is "future-compatible" not vice versa. If Gutenberg has some logic for the "runtime" script then maybe that's what I'm missing, but for now I'll just enqueue it myself and seems to work with or without GB.

**Comment by gziolo** [comment]
> @gziolo Hey did you see the comment: https://github.com/WordPress/gutenberg/commit/fcee058e3bb787ee041c170958e15e66a6302352#r150652268

No, I missed it.

> This fixes it:
>
> const fields = getBlockJsonScriptFields( blockJson ) || {};

Would you like to open PR with a bug fix proposed? You are 100% correct, this is one of the ways to address the case you explained when there are no scripts discovered.

> Also, is it normal that the *.asset.php files have 'handle' => 'undefined-[name]'?

No, it needs to get improved. Ideally, it should fallback to some other alternative way to get some more distinct name for the webpack build, or the name of the project defined in `package.json`.

> Another change in @wordpress/scripts that I must have missed (when upgrading from 30.5.1) is that I need to enqueue the runtime.js file that is generated to build dir myself? Or is there some logic that should do it automatically?

It should get enqueued automatically, but there might be some more edge cases as it's nontrivial.

> Is the Gutenberg part still the case? I usually develop without it and occasionally turn it on to check that what I have is "future-compatible" not vice versa. If Gutenberg has some logic for the "runtime" script then maybe that's what I'm missing, but for now I'll just enqueue it myself and seems to work with or without GB.

Gutenberg is no longer required. Where does this information come from? The logic for including the runtime is a mix of WordPress core code and the webpack build. First, webpack automatically injects the path to the runtime inside `block.json` file generated in the `build folder:

https://github.com/WordPress/gutenberg/blob/7a4b5aa5d8a63d97b084e1339e6ce544dfc944bb/packages/scripts/config/webpack.config.js#L344-L368

In fact, this is where the bug you mentioned needs to be fixed.

The code that registers the webpack runtime and other scripts defined for the block in `block.json` gets processed here:

https://github.com/WordPress/wordpress-develop/blob/12eaef1dd97ce46d7e665c69764cdae7387b03cc/src/wp-includes/blocks.php#L564-L593

**Pull request merged**